python/cudf/cudf/core/buffer/buffer.py

```diff
@@ -0,0 +1,298 @@
+# Copyright (c) 2020-2022, NVIDIA CORPORATION.
+
+from __future__ import annotations
+
+import math
+import pickle
+from types import SimpleNamespace
+from typing import Any, Dict, Mapping, Sequence, Tuple, Type, TypeVar
+
+import numpy
+
+import rmm
+
+import cudf
+from cudf.core.abc import Serializable
+from cudf.utils.string import format_bytes
+
+T = TypeVar("T", bound="Buffer")
+
+
+def cuda_array_interface_wrapper(ptr: int, size: int, owner: object = None):
+    """Wrap device pointer in an object that exposes `__cuda_array_interface__`
+
+    Parameters
+    ----------
+    ptr : int
+        An integer representing a pointer to device memory.
+    size : int, optional
+        Size of device memory in bytes.
+    owner : object, optional
+        Python object to which the lifetime of the memory allocation is tied.
+        A reference to this object is kept in the returned wrapper object.
+
+    Return
+    ------
+    SimpleNamespace
+        An object that exposes `__cuda_array_interface__` and keeps a reference
+        to `owner`.
+    """
+
+    if size < 0:
+        raise ValueError("size cannot be negative")
+
+    return SimpleNamespace(
+        __cuda_array_interface__={
+            "data": (ptr, False),
+            "shape": (size,),
+            "strides": None,
+            "typestr": "|u1",
+            "version": 0,
+        },
+        owner=owner,
+    )
+
+
+class Buffer(Serializable):
+    """A Buffer represents device memory.
+
+    Use the factory function `as_buffer` to create a Buffer instance.
+    """
+
+    _ptr: int
+    _size: int
+    _owner: object
+
+    def __init__(self):
+        raise ValueError(
+            f"do not create a {self.__class__} directly, please "
+            "use the factory function `cudf.core.buffer.as_buffer`"
+        )
+
+    @classmethod
+    def _from_device_memory(cls: Type[T], data: Any) -> T:
+        """Create a Buffer from an object exposing `__cuda_array_interface__`.
+
+        No data is being copied.
+
+        Parameters
+        ----------
+        data : device-buffer-like
+            An object implementing the CUDA Array Interface.
+
+        Returns
+        -------
+        Buffer
+            Buffer representing the same device memory as `data`
+        """
+
+        # Bypass `__init__` and initialize attributes manually
+        ret = cls.__new__(cls)
+        ret._owner = data
```

Comment:
If `data` is a `Buffer`, do we just want to grab its `_owner` and keep a reference to that? Or does doing so break our spilling logic somehow?

```suggestion
        if isinstance(data, Buffer):
            ret._owner = data._owner
        else:
            ret._owner = data
```

Reply:
It will not break the spilling logic but `data` should never be a `Buffer` instance. `as_buffer()` return as-is if the inpurt is a `Buffer`.